In [1]:
import pandas as pd

car_data = pd.read_csv("vehicles_us.csv")
car_data.head()


FileNotFoundError: [Errno 2] No such file or directory: 'vehicles_us.csv'

In [3]:
import pandas as pd

car_data = pd.read_csv("../vehicles_us.csv")
car_data.head()


price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28

In [2]:
import pandas as pd

car_data = pd.read_csv("../vehicles_us.csv")
car_data.head()


EmptyDataError: No columns to parse from file

In [4]:
import os
from pathlib import Path

print("CWD:", os.getcwd())
print("Aquí:", os.listdir(".")[:30])
print("Arriba:", os.listdir("..")[:30])


CWD: c:\Users\VeronicaOstos\OneDrive\Cursos VOG\Tripleten\Proyecto Sprint 7\Notebooks
Aquí: ['EDA.ipynb', 'EDA_malo.txt']
Arriba: ['.ipynb_checkpoints', '.pytest_cache', '.vscode', 'age_group.py', 'app.py', 'archivo.py', 'archivo.txt', 'Notebooks', 'vehicles_us.csv', '__pycache__']


In [5]:
from pathlib import Path

candidates = list(Path.cwd().rglob("vehicles_us.csv"))
[(str(p), p.stat().st_size) for p in candidates]


[]

In [6]:
good_path = r"C:\Users\VeronicaOstos\OneDrive\Cursos VOG\Tripleten\Proyecto Sprint 7\vehicles_us.csv"
car_data = pd.read_csv(good_path)
car_data.head()


price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28

In [7]:
car_data.shape
car_data.head()


price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28

In [8]:
import pandas as pd

car_data = pd.read_csv("vehicles_us.csv")

car_data.shape, car_data.columns


FileNotFoundError: [Errno 2] No such file or directory: 'vehicles_us.csv'

In [1]:
import pandas as pd

car_data = pd.read_csv("../vehicles_us.csv")


FileNotFoundError: [Errno 2] No such file or directory: '../vehicles_us.csv'

In [2]:
import os
os.getcwd()


'c:\\Users\\VeronicaOstos\\OneDrive\\Cursos VOG\\Tripleten\\Proyecto Sprint 7'

In [3]:
import pandas as pd

car_data = pd.read_csv("vehicles_us.csv")
car_data.shape, car_data.columns


((51525, 13),
 Index(['price', 'model_year', 'model', 'condition', 'cylinders', 'fuel',
        'odometer', 'transmission', 'type', 'paint_color', 'is_4wd',
        'date_posted', 'days_listed'],
       dtype='object'))

In [4]:
car_data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


In [5]:
car_data.head()


price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28

In [6]:
car_data.isna().sum().sort_values(ascending=False)


is_4wd          25953
paint_color      9267
odometer         7892
cylinders        5260
model_year       3619
condition           0
model               0
price               0
fuel                0
type                0
transmission        0
date_posted         0
days_listed         0
dtype: int64

In [7]:
car_data.describe(include="all")


price    model_year       model  condition     cylinders  \
count    51525.000000  47906.000000       51525      51525  46265.000000   
unique            NaN           NaN         100          6           NaN   
top               NaN           NaN  ford f-150  excellent           NaN   
freq              NaN           NaN        2796      24773           NaN   
mean     12132.464920   2009.750470         NaN        NaN      6.125235   
std      10040.803015      6.282065         NaN        NaN      1.660360   
min          1.000000   1908.000000         NaN        NaN      3.000000   
25%       5000.000000   2006.000000         NaN        NaN      4.000000   
50%       9000.000000   2011.000000         NaN        NaN      6.000000   
75%      16839.000000   2014.000000         NaN        NaN      8.000000   
max     375000.000000   2019.000000         NaN        NaN     12.000000   

         fuel       odometer transmission   type paint_color   is_4wd  \
count   51525   43633.000000        51525  51525       42258  25572.0   
unique      5            NaN            3     13          12      NaN   
top       gas            NaN    automatic    SUV       white      NaN   
freq    47288            NaN        46902  12405       10029      NaN   
mean      NaN  115553.461738          NaN    NaN         NaN      1.0   
std       NaN   65094.611341          NaN    NaN         NaN      0.0   
min       NaN       0.000000          NaN    NaN         NaN      1.0   
25%       NaN   70000.000000          NaN    NaN         NaN      1.0   
50%       NaN  113000.000000          NaN    NaN         NaN      1.0   
75%       NaN  155000.000000          NaN    NaN         NaN      1.0   
max       NaN  990000.000000          NaN    NaN         NaN      1.0   

       date_posted  days_listed  
count        51525  51525.00000  
unique         354          NaN  
top     2019-03-17          NaN  
freq           186          NaN  
mean           NaN     39.55476  
std            NaN     28.20427  
min            NaN      0.00000  
25%            NaN     19.00000  
50%            NaN     33.00000  
75%            NaN     53.00000  
max            NaN    271.00000

In [8]:
car_data.isna().mean().sort_values(ascending=False)


is_4wd          0.503697
paint_color     0.179854
odometer        0.153168
cylinders       0.102086
model_year      0.070238
condition       0.000000
model           0.000000
price           0.000000
fuel            0.000000
type            0.000000
transmission    0.000000
date_posted     0.000000
days_listed     0.000000
dtype: float64

In [9]:
car_data['date_posted'] = pd.to_datetime(car_data['date_posted'], errors='coerce')


In [10]:
car_data['car_age'] = car_data['date_posted'].dt.year - car_data['model_year']


In [11]:
car_data[car_data['car_age'] < 0].head()


Empty DataFrame
Columns: [price, model_year, model, condition, cylinders, fuel, odometer, transmission, type, paint_color, is_4wd, date_posted, days_listed, car_age]
Index: []

In [12]:
car_data['model_year'].describe()


count    47906.000000
mean      2009.750470
std          6.282065
min       1908.000000
25%       2006.000000
50%       2011.000000
75%       2014.000000
max       2019.000000
Name: model_year, dtype: float64

In [13]:
car_data['model_year'] = car_data['model_year'].fillna(car_data['model_year'].median())


In [14]:
car_data['date_posted'] = pd.to_datetime(car_data['date_posted'])


In [15]:
car_data['car_age'] = car_data['date_posted'].dt.year - car_data['model_year']


In [16]:
car_data[['model_year', 'date_posted', 'car_age']].head()


model_year date_posted  car_age
0      2011.0  2018-06-23      7.0
1      2011.0  2018-10-19      7.0
2      2013.0  2019-02-07      6.0
3      2003.0  2019-03-22     16.0
4      2017.0  2019-04-02      2.0

In [17]:
car_data['car_age'].describe()


count    51525.000000
mean         8.469229
std          6.079773
min          0.000000
25%          4.000000
50%          7.000000
75%         12.000000
max        110.000000
Name: car_age, dtype: float64

In [18]:
car_data['car_age'].describe()


count    51525.000000
mean         8.469229
std          6.079773
min          0.000000
25%          4.000000
50%          7.000000
75%         12.000000
max        110.000000
Name: car_age, dtype: float64

### Outliers en la variable `car_age`

Durante el análisis exploratorio se identificaron valores atípicos en la variable `car_age`, con una edad máxima de 110 años.  
Aunque este valor se encuentra fuera del rango típico de uso para la mayoría de los vehículos, no se considera un error de datos.

Este tipo de observaciones puede corresponder a autos clásicos o de colección que aún se encuentran en venta en el mercado.  
Por esta razón, los valores atípicos se conservan en el conjunto de datos para evitar la eliminación de información potencialmente relevante y mantener la diversidad del mercado representado.

“La conservación de estos valores permite un análisis más representativo del mercado real de vehículos usados.”

In [19]:
car_data['odometer'].describe()


count     43633.000000
mean     115553.461738
std       65094.611341
min           0.000000
25%       70000.000000
50%      113000.000000
75%      155000.000000
max      990000.000000
Name: odometer, dtype: float64

In [20]:
median_odometer = car_data['odometer'].median()
car_data['odometer'] = car_data['odometer'].fillna(median_odometer)


In [21]:
car_data['odometer'].isna().sum()


np.int64(0)

### Valores faltantes en la variable `odometer`

La variable `odometer` presenta aproximadamente un 15% de valores faltantes.
Dado que la distribución del kilometraje es asimétrica y contiene valores
extremos elevados, se decidió imputar los valores faltantes utilizando la
mediana, ya que es una medida robusta frente a outliers.

Esta decisión permite conservar la mayor cantidad de observaciones sin
distorsionar la distribución original de la variable.
supongo que:

0 km → posible error o auto nuevo

990,000 km → vehículos comerciales / trabajo

In [22]:
car_data['paint_color'] = car_data['paint_color'].fillna('unknown')


In [23]:
car_data['paint_color'].isna().sum()


np.int64(0)

### Valores faltantes en la variable `paint_color`

La variable categórica `paint_color` presenta aproximadamente un 18% de valores
faltantes. Dado que el color no puede inferirse de forma confiable y que eliminar
las observaciones implicaría una pérdida significativa de información, se optó
por crear una categoría explícita denominada `unknown`.

Esta estrategia permite conservar todas las observaciones y evita introducir
sesgo al imputar un valor arbitrario.


In [24]:
car_data['is_4wd'] = car_data['is_4wd'].fillna(0)


In [25]:
car_data['is_4wd'].value_counts()


is_4wd
0.0    25953
1.0    25572
Name: count, dtype: int64

### Tratamiento de valores faltantes en `is_4wd`

La variable `is_4wd` indica si el vehículo cuenta con tracción en las cuatro
ruedas (1 = sí). Los valores faltantes se interpretan como ausencia de esta
característica, práctica común en listados de vehículos donde solo se marca
cuando el atributo está presente.

Por esta razón, los valores nulos se imputaron con 0, indicando que el vehículo
no es 4WD.


In [26]:
car_data.loc[car_data['odometer'] == 0, 'odometer'] = None


In [27]:
median_odometer = car_data['odometer'].median()
car_data['odometer'] = car_data['odometer'].fillna(median_odometer)


In [28]:
car_data['odometer'].describe()


count     51525.000000
mean     115568.077555
std       59509.266582
min           1.000000
25%       80000.000000
50%      113000.000000
75%      146541.000000
max      990000.000000
Name: odometer, dtype: float64

### Tratamiento de valores atípicos y faltantes en `odometer`

La variable `odometer` presenta valores iguales a 0, los cuales no son
realistas para vehículos usados y se interpretan como datos faltantes.
Estos valores fueron reemplazados por NaN.

Posteriormente, los valores faltantes se imputaron utilizando la mediana,
debido a la distribución sesgada del kilometraje y a la presencia de valores
extremos.

Los valores elevados de kilometraje no se eliminaron, ya que pueden
corresponder a vehículos de uso intensivo o comercial y forman parte
legítima del mercado analizado.


In [29]:
car_data['cylinders'].mode()


0    8.0
Name: cylinders, dtype: float64

In [30]:
mode_cylinders = car_data['cylinders'].mode()[0]


In [31]:
car_data['cylinders'] = car_data['cylinders'].fillna(mode_cylinders)


In [32]:
car_data['cylinders'].describe()


count    51525.000000
mean         6.316623
std          1.672586
min          3.000000
25%          4.000000
50%          6.000000
75%          8.000000
max         12.000000
Name: cylinders, dtype: float64

### Tratamiento de valores faltantes en `cylinders`

La variable `cylinders` representa una característica técnica discreta
del vehículo. Los valores faltantes fueron imputados utilizando la moda,
ya que es la opción más representativa para variables categóricas numéricas
y evita introducir valores artificiales.

No se eliminaron observaciones para conservar la integridad del conjunto
de datos.


In [33]:
car_data['paint_color'].isna().mean()


np.float64(0.0)

In [34]:
car_data['paint_color'] = car_data['paint_color'].fillna('unknown')


In [35]:
car_data['paint_color'].value_counts()


paint_color
white      10029
unknown     9267
black       7692
silver      6244
grey        5037
blue        4475
red         4421
green       1396
brown       1223
custom      1153
yellow       255
orange       231
purple       102
Name: count, dtype: int64

### Tratamiento de valores faltantes en `paint_color`

La variable `paint_color` es de tipo categórico y presenta una proporción
relevante de valores faltantes. Dado que el color no es una característica
crítica para el análisis del comportamiento del mercado ni del precio,
los valores nulos fueron imputados con la categoría `"unknown"`.

Esta decisión permite conservar todos los registros sin introducir
suposiciones incorrectas sobre el color real del vehículo.


In [36]:
car_data['is_4wd'].value_counts(dropna=False)


is_4wd
0.0    25953
1.0    25572
Name: count, dtype: int64

In [37]:
car_data['is_4wd'] = car_data['is_4wd'].fillna(0)


In [38]:
car_data['is_4wd'] = car_data['is_4wd'].astype(int)


In [39]:
car_data['is_4wd'].value_counts()


is_4wd
0    25953
1    25572
Name: count, dtype: int64

### Tratamiento de valores faltantes en `is_4wd`

La variable `is_4wd` indica si el vehículo cuenta con tracción en las cuatro ruedas.
En este conjunto de datos, los valores faltantes representan vehículos que no
tienen esta característica.

Por lo tanto, los valores nulos fueron imputados con `0`, y la variable fue
convertida a tipo entero para facilitar su análisis como variable binaria.


In [40]:
car_data['condition'].value_counts()
car_data['fuel'].value_counts()
car_data['transmission'].value_counts()
car_data['type'].value_counts()


type
SUV            12405
truck          12353
sedan          12154
pickup          6988
coupe           2303
wagon           1541
mini-van        1161
hatchback       1047
van              633
convertible      446
other            256
offroad          214
bus               24
Name: count, dtype: int64

In [41]:
car_data['date_posted'].dtype


dtype('<M8[ns]')

In [42]:
car_data['date_posted'] = pd.to_datetime(car_data['date_posted'])


In [43]:
car_data['car_age'] = car_data['date_posted'].dt.year - car_data['model_year']


In [44]:
car_data[['model_year', 'date_posted', 'car_age']].head()


model_year date_posted  car_age
0      2011.0  2018-06-23      7.0
1      2011.0  2018-10-19      7.0
2      2013.0  2019-02-07      6.0
3      2003.0  2019-03-22     16.0
4      2017.0  2019-04-02      2.0

### Creación de la variable `car_age`

Se creó la variable `car_age` como la diferencia entre el año de publicación
del anuncio (`date_posted`) y el año del modelo del vehículo (`model_year`).
Esta variable permite analizar el efecto de la antigüedad del vehículo sobre
el precio de venta.
